src/deadline/nuke_adaptor/copycat_adaptor.py

```diff
@@ -0,0 +1,143 @@
+import argparse
+import json
+import logging
+import pathlib
+import re
+import subprocess
+import sys
+import threading
+import time
+from typing import Dict, List, Optional
+
+error_regex = re.compile(r".*ERROR: (.+)")
+step_complete_regex = re.compile(r"\[Step:(\d+)/(\d+)\].*")
+
+_current_percent_done = 0
+
+
+def _report_progress(current_step, total_steps):
+    global _current_percent_done
+    # round progress to a single decimal point
+    progress = round(current_step * 100.0 / total_steps, 1)
+    if progress != _current_percent_done:
+        _current_percent_done = progress
+        return f"openjd_progress: {progress}"
+    else:
+        return None
+
+
+_error_encountered = False
+
+
+# returning message rather than printing to make this unit testable
+def report_openjd_messages(line) -> Optional[str]:
+    global _error_encountered
+    if (match := error_regex.match(line)) is not None:
+        # error message is captured in group(1)
```

Comment:
where was this group(1) defined? I think you are referring to regex group but the comment might not help that much.